commonmark-ext-metadata/src/main/java/org/commonmark/ext/metadata/internal/MetadataBlockParser.java

```diff
@@ -0,0 +1,147 @@
+                    node.setKey(key);
+                    node.setValues(values);
+                    block.appendChild(node);
+                }
+
+                literal = false;
+                key = matcher.group(1);
+                values = new ArrayList<>();
+                if ("|".equals(matcher.group(2))) {
+                    literal = true;
+                } else if (!"".equals(matcher.group(2))) {
+                    values.add(matcher.group(2));
+                }
+            } else {
+                if (literal) {
+                    matcher = REGEX_METADATA_LITERAL.matcher(line);
+                    if (matcher.matches()) {
+                        if (values.size() == 1) {
+                            values.set(0, values.get(0) + "\n" + matcher.group(1).trim());
+                        } else {
+                            values.add(matcher.group(1).trim());
+                        }
+                    }
+                } else {
+                    matcher = REGEX_METADATA_LIST.matcher(line);
+                    if (matcher.matches()) {
+                        values.add(matcher.group(1));
+                    }
+                }
+            }
+        }
+
+        if (key != null) {
+            MetadataNode node = new MetadataNode();
+            node.setKey(key);
+            node.setValues(values);
+            block.appendChild(node);
+        }
+    }
+
+    public static class Factory extends AbstractBlockParserFactory {
+        @Override
+        public BlockStart tryStart(ParserState state, MatchedBlockParser matchedBlockParser) {
+            CharSequence line = state.getLine();
+            BlockParser parentParser = matchedBlockParser.getMatchedBlockParser();
+            // check whether this line is the first line of whole document or not
+            if (parentParser instanceof DocumentBlockParser && parentParser.getBlock().getFirstChild() == null &&
+                    REGEX_BOUNDARY.matcher(line).matches()) {
+                // count valid metadata line in metadata block
+                int prevIndex;
+                int index = nextLineEnd(line, 0);
+                int validLineCount = 0;
+                CharSequence subseq;
+                do {
+                    prevIndex = index + 1;
+                    index = nextLineEnd(line, prevIndex);
+                    subseq = line.subSequence(prevIndex, index);
+                    if (REGEX_METADATA.matcher(subseq).matches()) {
+                        validLineCount++;
+                    }
+                } while (!REGEX_BOUNDARY.matcher(subseq).matches() && prevIndex != index);
+
+                if (validLineCount > 0) {
+                    return BlockStart.of(new MetadataBlockParser()).atIndex(state.getNextNonSpaceIndex());
+                }
+            }
+
+            return BlockStart.none();
+        }
+
+        private int nextLineEnd(CharSequence seq, int startIndex) {
```

Comment:
Thanks for investigating!

Looks like the Python library implements it as a pre-processor that removes the metadata block from the input. Then later the rest of the document is parsed. I'm a bit hesitant to explore that direction because it makes source positions more complicated. Also, that doesn't really work nicely when parsing from a Reader instead of a String.

So let's go with keeping it simple for now, and if the first line in the document happens to be a `---`, we interpret it as a metadata block. We could do some more checking in `tryContinue` and stop the block if the first line of content doesn't look like metadata content.

That means we'll need to override example 14 and 16 in `SpecIntegrationTest`, see `getOverriddenExamples`.

Also, +1 for only allowing `---`.